src/booster.rs

```diff
@@ -274,29 +274,72 @@ impl Booster {
     /// Get names of evaluation metrics
     pub fn eval_names(&self) -> Result<Vec<String>> {
         let num_metrics = self.num_eval()?;
-        let metric_name_length = 32;
+
+        /////////////////////////////////////////////////////////////////////
+        // call with 0-sized buffer to find out how much space to allocate
+        /////////////////////////////////////////////////////////////////////
         let mut num_eval_names = 0;
         let mut out_buffer_len = 0;
-        let out_strs = (0..num_metrics)
-            .map(|_| {
-                CString::new(" ".repeat(metric_name_length))
-                    .unwrap()
-                    .into_raw() as *mut c_char
-            })
+
+        lgbm_call!(lightgbm_sys::LGBM_BoosterGetEvalNames(
```

Comment:
Does this work correctly if the 2nd and 4th parameters are zero? According to the docs, these are input parameters specifiying the number of char* an out_strs and the size of the pre-allocated strings.

Reply:
What I'm claiming is "I have a preallocated buffer that can hold 0 strings, each 0 bytes long". That buffer happens to be null-pointer. But you are allowed to write exactly 0 bytes to a null-pointer, so that's fine unless lgbm overflows the buffer.

The API documentation is a bit spotty, but the only valid ways lgbm can handle that call without overflowing the buffer is either handing me back an error code, or filling the output variables without filling the buffer and returning success. It does the latter, and I get the value of out_buffer_len to know how much to actually allocate.

The alternative is allocating some arbitrary buffer size, and if it turns out to be too small either live with truncated names, or allocate a larger buffer and try again. This version is a bit slower (always calling twice) but simpler and without truncation.